tests/test_date.py

```diff
@@ -132,76 +205,180 @@ def test_periods(self, low, high, period, expected_results):
+        self.period = period
+
+    def when_date_intersecting_generated(self):
+        self.intersected_dates = self.dates
+
+    def when_should_one_date_for_each_month(self):
+        self.intersected_dates = list(date.get_intersecting_periods(
+            self.period_low, self.period_high, period=self.period))
+
+    def when_empty_period_is_parsed(self):
+        self.intersected_dates = list(date.get_intersecting_periods(
+            self.period_low, self.period_high, period=self.period))
+
+    def then_should_one_date_for_each_month(self):
+        self.assertEquals(self.expected_results, self.intersected_dates)
+
+    def then_date_range_length_should_one(self):
+        self.assertEquals(1, len(self.intersected_dates))
+
+    def then_date_period_and_intersected_dates_should_equal(self):
+        self.assertEquals(self.period_low, self.intersected_dates[0])
+
+    def then_period_is_empty(self):
+        self.assertEquals([], self.intersected_dates)
+
+    def then_should_reject_easily_mistaken_dateutil_arguments(self):
+        with self.assertRaisesRegexp(ValueError, "Invalid period: {}".format(self.invalid_period)):
+            date.get_intersecting_periods(self.period_low, self.period, period=self.invalid_period).next()
+
+
+class ParseDateWithFormats(BaseTestCase):
+    def setUp(self):
+        super(ParseDateWithFormats, self).setUp()
+        self.dt_data = NotImplemented
+
+    def test_date_with_not_matching_format_is_not_parsed(self):
+        self.given_invalid_date('yesterday')
+        self.when_parse_invalid_date()
+        self.then_shouldnt_parse_invalid_date()
 
     def test_should_parse_date(self):
-        result = date.parse_with_formats('25-03-14', ['%d-%m-%y'])
-        self.assertEquals(datetime(2014, 3, 25).date(), result['date_obj'].date())
+        self.given_date_str('25-03-14')
+        self.given_date_format('%d-%m-%y')
+        self.given_date(datetime(2014, 3, 25))
+        self.when_should_parse_date()
+        self.then_should_parse_date()
 
     def test_should_use_current_year_for_dates_without_year(self):
-        today = datetime.today()
+        self.given_current_year()
+        self.when_should_use_current_year_for_dates_without_year()
+        self.then_should_use_current_year_for_dates_without_year()
+
+    @parameterized.expand([
+        param(given_date=datetime(2014, 8, 31),
+              mock_date=datetime(2014, 8, 12),
+              date_str='August 2014',
+              date_format='%B %Y'),
+        param(given_date=datetime(2014, 2, 28),
+              mock_date=datetime(2014, 2, 12),
+              date_str='February 2014',
+              date_format='%B %Y'),
+    ])
+    def test_should_use_current_date_for_dates_without_day(self, given_date, mock_date,
+                                                           date_str, date_format):
+        self.given_date(given_date)
+        self.given_datetime_mock(mock_date)
+        self.given_patch()
```

Comment:
But we are patching [datetime.utcnow](https://docs.python.org/2/library/datetime.html#datetime.datetime.utcnow), Don't you think it should be mentioned in name?
Think of `test_*` methods body as of a text that would be read by managers without any technical knowledge. They don't really care what we do to achieve required goal, they only care what code does. So, considering the above, 'patch' in name of `given_*`, `when_*`, `then_*` names is bad, and 'now' is good.

We can of course use 'patch' and like in other methods (like we have `add_patch` method), but we should try to keep our `test_*` methods technical-knowledge-free :smile: 

p.s. In general it is a good rule to follow that callables should be named so that it is clear what they do without looking into code itself.